File: LLD/Design-Patterns/Singleton/code/singleton.py
"""
Singleton Pattern — implemented with a metaclass so any class using
SingletonMeta automatically becomes a singleton.
Run: python singleton.py
"""

import threading


class SingletonMeta(type):
    _instances = {}
    _lock = threading.Lock()

    def __call__(cls, *args, **kwargs):
        # double-checked locking, same idea as the Java version
        if cls not in cls._instances:
            with cls._lock:
                if cls not in cls._instances:
                    cls._instances[cls] = super().__call__(*args, **kwargs)
        return cls._instances[cls]


class Singleton(metaclass=SingletonMeta):
    def __init__(self):
        self.request_count = 0
        print("Singleton instance created.")

    def log_request(self):
        self.request_count += 1
        print(f"Handled request #{self.request_count}")


if __name__ == "__main__":
    a = Singleton()
    b = Singleton()

    a.log_request()
    b.log_request()

    print("a is b:", a is b)  # True -> same instance

File: LLD/Design-Patterns/Singleton/code/singleton.js
/**
 * Singleton Pattern — JavaScript's module cache already gives you a
 * singleton for free (require/import only evaluates a module once),
 * but this shows the classic explicit pattern too, using a class with
 * a static instance and a frozen export.
 *
 * Run: node singleton.js
 */

class Singleton {
  static #instance; // private static field
  #requestCount = 0;

  constructor() {
    if (Singleton.#instance) {
      throw new Error("Use Singleton.getInstance() instead of `new Singleton()`");
    }
    console.log("Singleton instance created.");
  }

  static getInstance() {
    if (!Singleton.#instance) {
      Singleton.#instance = new Singleton();
    }
    return Singleton.#instance;
  }

  logRequest() {
    this.#requestCount++;
    console.log(`Handled request #${this.#requestCount}`);
  }
}

const a = Singleton.getInstance();
const b = Singleton.getInstance();

a.logRequest();
b.logRequest();

console.log("a === b:", a === b); // true -> same instance

module.exports = Singleton;
